libmscore/rendermidi.cpp

```diff
@@ -731,6 +735,478 @@ bool Score::isSubdivided(ChordRest* chord, int swingUnit)
+// and excluding notes which are tied to a previous note.
+// The given function, visit, is caled on each note until it returns true, terminating the iteration.
+// If visit() never returns true, then finally elsevisit is called with no argument.
+void visitNotes(Segment *segment, Staff *staff, std::function<bool(Note *)> visit, std::function<void()> elsevisit) {
+   int staffIdx = staff->idx();
+   int startTrack = staffIdx * VOICES;
+   int endTrack   = startTrack + VOICES;
+   for (int track = startTrack; track < endTrack; ++track) {
+       Element *e = segment->element(track);
+       if (!e || e->type() != Element::Type::CHORD)
+            continue;
+       Chord* chord = static_cast<Chord*>(e);
+       for (Note* note : chord->notes()) {
+           if (note->tieBack())
+                continue;
+           if ( visit(note) )
+                return;
+        }
+        elsevisit();
+    }
+}
+
+// find the line in clefF corresponding to lineL2 in clefR.
+int convertLine (int lineL2, ClefType clefL, ClefType clefR) {
+   int lineR2 = lineL2;
+   int goalpitch = line2pitch(lineL2, clefL, Key::C);
+   while ( line2pitch(lineR2, clefR, Key::C) > goalpitch )
+       lineR2++;
+   while ( line2pitch(lineR2, clefR, Key::C) < goalpitch )
+       lineR2--;
+   return lineR2;
+};
+
+int convertLine(int lineL2, Note *noteL, Note *noteR) {
+    return convertLine(lineL2,
+                       noteL->chord()->staff()->clef(noteL->chord()->tick()),
+                       noteR->chord()->staff()->clef(noteR->chord()->tick()));
+}
+
+int articulationExcursion (Note *noteL, Note *noteR, int deltastep) {
+    // noteL is the note to measure the deltastep from, i.e., ornaments are w.r.t. this note
+    // noteR is the note to search backward from to find accidentals.
+    //    for ornament calcuation noteL and noteR are the same, but for glissando they are
+    //     the start end end note of glissando.
+    // deltastep is the number of diatonic steps between the base note and this articulation step.
+    if ( 0 == deltastep )
+        return 0;
+    Chord *chordL = noteL->chord();
+    Chord *chordR = noteR->chord();
+    int pitchL = noteL->pitch();
+    int tickL = chordL->tick();
+    Staff * staffL = chordL->staff();
+    ClefType clefL = staffL->clef(tickL);
+    // line represents the ledger line of the staff.  0 is the top line, 1, is the space between the top 2 lines,
+    //  ... 8 is the bottom line.
+    int lineL     = noteL->line();
+    // we use line - deltastep, because lines are oriented from top to bottom, while step is oriented from bottom to top.
+    int lineL2    = lineL - deltastep;
+    auto measureR = chordR->segment()->measure();
+    
+    Segment *S = noteL->chord()->segment();
+    int lineR2 = convertLine(lineL2, noteL, noteR);
+    // is there another note in this segment on the same line?
+    // if so, use its pitch exactly.
+    int halfsteps;
+    visitNotes( S, staffL,
```

Comment:
Ok, I can have another think about it.  and in the end you’re the boss so I’ll do what you recommend.
It may not be very difficult, since the code works now.  If I refactor and break something, I’ll know immediately
as the tests will fail.

C++11 has added some very nice features to the language to make is much easier to use, and specifically
more accessible to non-C++ experts such as myself.  In my as-of-yet limited experience, long time c++ programers
don’t like the new features, and find them difficult to read.  I think non-long-time-c++ users find them much
easier.





> On 18 May 2015, at 14:09, Nicolas Froment <notifications@github.com> wrote:
> 
> In libmscore/rendermidi.cpp <https://github.com/musescore/MuseScore/pull/1946#discussion_r30501507>:
> 
> > +    int tickL = chordL->tick();
> > +    Staff * staffL = chordL->staff();
> > +    ClefType clefL = staffL->clef(tickL);
> > +    // line represents the ledger line of the staff.  0 is the top line, 1, is the space between the top 2 lines,
> > +    //  ... 8 is the bottom line.
> > +    int lineL     = noteL->line();
> > +    // we use line - deltastep, because lines are oriented from top to bottom, while step is oriented from bottom to top.
> > +    int lineL2    = lineL - deltastep;
> > +    auto measureR = chordR->segment()->measure();
> > +    
> > +    Segment *S = noteL->chord()->segment();
> > +    int lineR2 = convertLine(lineL2, noteL, noteR);
> > +    // is there another note in this segment on the same line?
> > +    // if so, use its pitch exactly.
> > +    int halfsteps;
> > +    visitNotes( S, staffL,
> No. I think it wouldn't help. I understand where you come from (thus the LISP joke) I'm just trying to keep the MuseScore codebase along the same line to ease the long term maintenance.
> I saw the other local functions all over your code. If you check the rest of MuseScore codebase you will see that the use of local function is almost non existent. The use of function pointers is also very small. So I'm just concerned about adding "new ways" to write code that would make the codebase heterogeneous and so harder to maintain.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/1946/files#r30501507>.
>